packages/client/src/objectSet/ObjectSetListenerWebsocket.ts

```diff
@@ -151,59 +217,87 @@ export class ObjectSetListenerWebsocket {
       ]);
 
       // the consumer may have already unsubscribed before we are ready to request a subscription
-      if (!this.#listeners.has(requestId)) {
+      // so we have to acquire the pendingSubscription after the await.
+      if (sub.status === "done" || sub.status === "error") {
         return;
       }
 
-      // subscribe to object set
-      const subscribe: ObjectSetSubscribeRequests = {
-        id: requestId,
-        requests: [{
-          objectSet: temporaryObjectSet.objectSetRid,
-          objectSetContext: {
-            objectSetFilterContext: { parameterOverrides: {} },
-          },
-          watchAllLinks: false,
-        }],
-      };
+      // Use new temporary object set id
+      sub.temporaryObjectSetId = temporaryObjectSet.objectSetRid;
 
-      this.#ws?.send(JSON.stringify(subscribe));
+      this.#sendSubscribeMessage();
     } catch (error) {
-      this.#getCallbackByRequestId(requestId, "onError")?.(error);
+      sub.listener.onError(error);
+    }
+  }
+
+  #sendSubscribeMessage() {
+    // If two calls to `.subscribe()` happen at once (or if the connection is reset),
+    // we may have multiple subscriptions that don't have a subscriptionId yet,
+    // so we filter those out.
+    const readySubs = [...this.#subscriptions.values()].filter(isReady);
+    if (readySubs.length === 0) return;
+
+    const id = crypto.randomUUID();
```

Comment:
why do you do the fallback  in `subscribe` but not here?

Reply:
There are two types of ids at play. A requestId for initiating the subscription, chosen by the consumer of the api. The result is tied to this value. And gives an array of the second type: subscriptionId. There is one of these for each subscription entry but the server controls them.

Reply:
i understand the different IDs, but im asking about why you fallback to node crypto in the `subscribe`function  but not here when you call `crypto`

Reply:
Oh! Because it assigns to globalThis. It sure to be set after that as thats the public entry point and you need to call it before this method is ever called. Will add a comment